docs/proposal-new-output-syntax.md

```diff
@@ -0,0 +1,702 @@
+# Proposal: Optional Output Syntax
+
+Spawn-on-demand solves many problems but in its pure form it can't tell if the
+workflow has followed the "intended path" - to completion of the whole graph
+(say) vs going down a dead-end side branch. This is because all task outputs 
+are effectively *optional*, and the scheduler just follows the flow wherever it
+leads at runtime.
+
+Cylc 7 could tell (albeit imperfectly!) if the intended path was followed 
+because all task outputs were effectively *required*. All possible paths are
+pre-spawned (to the next cycle point instance of each active task, roughly
+speaking). Users had to explicitly remove unused alternate paths with suicide
+triggers, and any remaining unsatisfied waiting tasks would indicate a problem
+by stalling the workflow.
+
+This document proposes a way to convey the intended path of execution to the
+Cylc 8 spawn-on-demand scheduler by making all task outputs **expected** by
+default and giving a new syntax to explicitly mark **optional** outputs. If a
+task finishes without completing all of its expected outputs it will be
+retained in the n=0 task pool and flagged as an **incomplete task** (if the
+task did not fail unexpectedly then this represents an error against the
+workflow configuration that was not recognized by the task job itself).
+
+*Note: optional triggers were considered and rejected as another way of
+potentially handling this problem - see cylc-admin#128*
+
+## The New Output Syntax
+
+`foo:x` means `x` is an *expected output* of task `foo`. If a task finishes
```

Comment:
I'm a bit sleep befuddled/under-caffinated this morning but I'm finding this easier to understand if I `s/expected/required/`

Reply:
I considered "required" too, but it sounds more coercive. We're not actually doing anything to "require" that the output gets completed, but the graph config nevertheless "expects" it to be completed (and gets a surprise if it isn't).

Reply:
Actually I've changed my mind on this!  If an output is *required* it can't be *optional*, but if it is optional we can *expect* that it might not be completed.